presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -764,17 +764,21 @@ public int read(byte[] buffer, int offset, int length)
                 int bytesRead = retry()
                         .maxAttempts(maxAttempts)
                         .exponentialBackoff(new Duration(1, TimeUnit.SECONDS), maxBackoffTime, maxRetryTime, 2.0)
-                        .stopOn(InterruptedException.class, UnrecoverableS3OperationException.class)
+                        .stopOn(InterruptedException.class, UnrecoverableS3OperationException.class, AbortedException.class)
                         .onRetry(STATS::newReadRetry)
                         .run("readStream", () -> {
                             seekStream();
                             try {
                                 return in.read(buffer, offset, length);
                             }
-                            catch (Exception e) {
-                                STATS.newReadError(e);
+                            catch (AbortedException abortedException) {
```

Comment:
The `Error processing` messages for aborted tasks should be handled directly, probably by ignoring the error if the task is cancelled. (It would be impossible to fix all such cases at the edges like this)

You're right about EOF. I was thinking this was like `readFully()`. However, returning `0` actually seems to be invalid, per the docs for `read`:

> If len is zero, then no bytes are read and 0 is returned; otherwise, there is an attempt to read at least one byte. If no byte is available because the stream is at end of file, the value -1 is returned; otherwise, at least one byte is read and stored into b.

In general, I think it's a bad idea for code to make assumptions about how it's used. This is an `InputStream`, so it should follow that contract. Otherwise, it's easy to have it be used in an unexpected way and cause subtle bugs.

One existing problem that violates the expected contract is throwing non-`IOException`s like `AbortedException`, since code using an `InputStream` doesn't expect a `RuntimeException`.